server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java

```diff
@@ -87,9 +97,107 @@ public SearchResponseData load(Collector collector, @Nullable Facets facets) {
+    Set<String> issueKeysToLoad = copyOf(difference(ImmutableSet.copyOf(collector.getIssueKeys()), preloadedIssueKeys));
+    if (issueKeysToLoad.isEmpty()) {
+      return preloadedIssues;
+    }
+
+    List<IssueDto> loadedIssues = dbClient.issueDao().selectByKeys(dbSession, issueKeysToLoad);
+    return concat(preloadedIssues.stream(), loadedIssues.stream())
+      .collect(toList(preloadedIssues.size() + loadedIssues.size()));
+  }
+
+  private void loadUsers(SearchResponseData preloadedResponseData, Collector collector, DbSession dbSession, SearchResponseData result) {
+    if (collector.contains(USERS)) {
+      List<UserDto> preloadedUsers = firstNonNull(preloadedResponseData.getUsers(), emptyList());
+      Set<String> preloadedLogins = preloadedUsers.stream().map(UserDto::getLogin).collect(MoreCollectors.toSet(preloadedUsers.size()));
+      Set<String> requestedLogins = collector.get(USERS);
+      Set<String> loginsToLoad = copyOf(difference(requestedLogins, preloadedLogins));
+
+      if (loginsToLoad.isEmpty()) {
+        result.setUsers(preloadedUsers);
+      } else {
+        List<UserDto> loadedUsers = dbClient.userDao().selectByLogins(dbSession, loginsToLoad);
+        result.setUsers(concat(preloadedUsers.stream(), loadedUsers.stream()).collect(toList(preloadedUsers.size() + loadedUsers.size())));
+      }
+    }
+  }
+
+  private void loadComponents(SearchResponseData preloadedResponseData, Collector collector, DbSession dbSession, SearchResponseData result) {
+    Collection<ComponentDto> preloadedComponents = preloadedResponseData.getComponents();
+    Set<String> preloadedComponentUuids = preloadedComponents.stream().map(ComponentDto::uuid).collect(MoreCollectors.toSet(preloadedComponents.size()));
+    Set<String> componentUuidsToLoad = copyOf(difference(collector.getComponentUuids(), preloadedComponentUuids));
+
+    result.addComponents(preloadedComponents);
+    if (!componentUuidsToLoad.isEmpty()) {
+      result.addComponents(dbClient.componentDao().selectByUuids(dbSession, componentUuidsToLoad));
+    }
+
+    // always load components and projects, because some issue fields still relate to component ids/keys.
+    // They should be dropped but are kept for backward-compatibility (see SearchResponseFormat)
+    result.addComponents(dbClient.componentDao().selectSubProjectsByComponentUuids(dbSession, collector.getComponentUuids()));
+    addProjectUuids(collector, dbSession, result);
+  }
+
+  private void addProjectUuids(Collector collector, DbSession dbSession, SearchResponseData result) {
+    Collection<ComponentDto> loadedComponents = result.getComponents();
+    for (ComponentDto component : loadedComponents) {
+      collector.addProjectUuid(component.projectUuid());
+    }
+    Set<String> loadedProjectUuids = loadedComponents.stream().filter(cpt -> cpt.uuid().equals(cpt.projectUuid())).map(ComponentDto::uuid).collect(MoreCollectors.toSet());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "loadedComponents" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)